src/proto/Utxo.proto

```diff
@@ -0,0 +1,113 @@
+syntax = "proto3";
+
+package TW.Utxo.Proto;
+option java_package = "wallet.core.jni.proto";
+
+import "Common.proto";
+
+enum Error {
+    OK = 0;
+    Error_sighash_failed = 1;
+    Error_invalid_wpkh_script_pubkey = 2;
+    Error_invalid_leaf_hash = 3;
+    Error_invalid_sighash_type = 4;
+    Error_invalid_lock_time = 5;
+    Error_invalid_txid = 6;
+    Error_one_prevout_out_of_bound = 7;
+    Error_missing_sighash_method = 8;
+}
+
+message SigningInput {
+    // The protocol version, is currently expected to be 1 or 2 (BIP68)
+    int32 version = 2;
+
+    // Block height or timestamp indicating at what point transactions can be
+    // included in a block.
+    oneof lock_time {
+        uint32 blocks = 3;
+        uint32 seconds = 4;
+    }
+
+    // The inputs of the transaction to sign. If an individual input does not
+    // provide an index, then the index of the inserted order is used.
+    repeated TxIn inputs = 5;
+
+    repeated TxOut outputs = 6;
+}
+
+message TxIn {
+    // The referenced transaction hash.
+    bytes txid = 2;
+
+    // The index of the referenced output.
+    uint32 vout = 3;
+
+    // The sequence number (TODO).
+    uint32 sequence = 7;
+
+    oneof sighash_method {
```

Comment:
Noob question. Shouldn't we also have `P2SH` here?

Reply:
This method (which I now renamed to `signing_method`) is only from the perspective of the Utxo structure which is not aware of any transaction standards (such as P2PKH, P2WPKH, etc). That field only tells this crate how to sign the input, such as "legacy" for P2SH/P2PKH, "segwit" for P2WPKH, etc. The Bitcoin input structure will look like this:

```protobuf
message Input {
	bytes txid = 1;

	uint32 vout = 2;

	uint64 amount = 3;

	Utxo.Proto.SighashType sighash_type = 4;

	oneof to_recipient {
		Builder builder = 5;
		ScriptWitness custom = 6;
	}

	bool one_prevout = 7;

	message Builder {
		oneof variant {
			// Pay-to-Script-Hash, specify the hash.
			bytes p2sh = 1;
			// Pay-to-Public-Key-Hash
			bytes p2pkh = 2;
			// Pay-to-Witness-Script-Hash, specify the hash.
			bytes p2wsh = 3;
			// Pay-to-Public-Key-Hash
			bytes p2wpkh = 6;
			// Pay-to-Taproot-key-path (balance transfers), specify the public key.
			bytes p2tr_key_path = 7;
			// Pay-to-Taproot-script-path (complex transfers)
			TaprootScriptPath p2tr_script_path = 8;
			Brc20Inscription brc20_inscribe = 9;
		}
	}

	message ScriptWitness {
		bytes script_sig = 1;
		repeated bytes witness_items = 2;
	}

	message TaprootScriptPath {
		bytes payload = 1;
		bytes control_block = 2;
	}

	message Brc20Inscription {
		bytes inscribe_to = 1;
		string ticker = 2;
		uint64 transfer_amount = 3;
		bytes control_block = 4;
	}
}
```

I will update everything of Friday evening.